core/src/container.cpp

```diff
@@ -806,46 +814,144 @@ void Alternatives::onNewSolution(const SolutionBase& s) {
+	liftSolution(s);
+}
+
+FallbacksPrivate::FallbacksPrivate(Fallbacks* me, const std::string& name)
+   : ParallelContainerBasePrivate(me, name) {}
+
+void FallbacksPrivate::initializeExternalInterfaces(InterfaceFlags expected) {
+	if (expected & READS_START)
+		starts().reset(new Interface([this](Interface::iterator external, bool updated) {
+		   this->onNewExternalState<Interface::FORWARD>(external, updated);
+		}));
+	if (expected & READS_END)
+		ends().reset(new Interface([this](Interface::iterator external, bool updated) {
+		   this->onNewExternalState<Interface::BACKWARD>(external, updated);
+		}));
+
+	// we've got to set this somewhere once the interface flags are known.
+	// so we might as well do it here
+	if(expected == GENERATE)
+		current_generator_ = children().begin();
+}
+
+void FallbacksPrivate::onNewFailure(const Stage& child, const InterfaceState* from, const InterfaceState* to) {
+	// only react to failure if it's the last possible candidate failing
+	// otherwise there might still be a feasible solution
+	if(&child == &*children().back())
+		ContainerBasePrivate::onNewFailure(child, from, to);
+}
+
+void FallbacksPrivate::computeGenerate() {
+	if(solutions_.empty())
+		// move to first generator that can run
+		while(current_generator_ != children().end() && !(*current_generator_)->pimpl()->canCompute())
+			++current_generator_;
+
+	if(current_generator_ == children().end())
 		return;
 
 	try {
-		active_child_->pimpl()->runCompute();
+		// run ALL possible computations (on new state)
+		// this is needed to decide whether it should be passed to the next child too
+		while ((*current_generator_)->pimpl()->canCompute()){
+			(*current_generator_)->pimpl()->runCompute();
+		}
 	} catch (const Property::error& e) {
-		active_child_->reportPropertyError(e);
+		(*current_generator_)->reportPropertyError(e);
 	}
 }
 
-void Fallbacks::onNewSolution(const SolutionBase& s) {
-	liftSolution(s);
+template <typename Interface::Direction dir>
+void FallbacksPrivate::onNewExternalState(Interface::iterator external, bool updated) {
+	// TODO(v4hn): updated is not implemented
+	if(updated){
+		ROS_DEBUG_NAMED("Fallback", "updating external states is not supported in Fallbacks");
+		return;
+	}
+
+	pending_states_.push_back(ExternalState(external, dir, children().begin()));
+}
+
+void FallbacksPrivate::computeFromExternal(){
+	assert(!pending_states_.empty());
+	auto spec { pending_states_.front() };
+
+	pending_states_.pop_front();
+
+	ROS_DEBUG_STREAM_NAMED("Fallback", "Push external state to '" << (*spec.stage)->name() << "'");
+	// feed a new state
+	copyState(spec.dir,
+	          spec.external_state,
+	          (*spec.stage)->pimpl()->pullInterface(spec.dir),
+	          false);
+
+	const auto& stage { (*spec.stage)->pimpl() };
+
+	try {
+		// run ALL possible computations (on new state)
+		// this is needed to decide whether it should be passed to the next child too
+		while (stage->canCompute()){
+			stage->runCompute();
```

Comment:
Same here: We should never run an exhaustive search on a single stage only.
Similar to generators, you could just keep track of the active child for each external states and only forward to the next child, if it returns `canCompute() = false`.

Reply:
I did not bother to include the logic for this before because it is impossible with the current interfaces to get rid of the exhaustive search altogether. Once we pass an external state to a child, Fallbacks *has* to commit to run that child until it's exhausted. I just pushed [an implementation to run only a single job per compute](https://github.com/ros-planning/moveit_task_constructor/pull/287/commits/2bac879c1bcc6946fc8cc874c61e22597a75bd95) nonetheless, because `Task` depends on `compute()` to return periodically to publish introspection.